src/libsystemd/sd-login/sd-login.c

```diff
@@ -84,155 +84,38 @@ _public_ int sd_pid_get_cgroup(pid_t pid, char **ret_cgroup) {
-        int r;
-
-        assert_return(pidfd >= 0, -EBADF);
-        assert_return(ret_slice, -EINVAL);
-
-        r = pidfd_get_pid(pidfd, &pid);
-        if (r < 0)
-                return r;
-
-        r = sd_pid_get_slice(pid, &slice);
-        if (r < 0)
-                return r;
-
-        r = pidfd_verify_pid(pidfd, pid);
-        if (r < 0)
-                return r;
-
-        *ret_slice = TAKE_PTR(slice);
-
-        return 0;
-}
-
-_public_ int sd_pidfd_get_user_slice(int pidfd, char **ret_slice) {
-        _cleanup_free_ char *slice = NULL;
-        pid_t pid;
-        int r;
-
-        assert_return(pidfd >= 0, -EBADF);
-        assert_return(ret_slice, -EINVAL);
-
-        r = pidfd_get_pid(pidfd, &pid);
-        if (r < 0)
-                return r;
-
-        r = sd_pid_get_user_slice(pid, &slice);
-        if (r < 0)
-                return r;
-
-        r = pidfd_verify_pid(pidfd, pid);
-        if (r < 0)
-                return r;
-
-        *ret_slice = TAKE_PTR(slice);
+#define DEFINE_PIDFD_GETTER(name)                                       \
+        _public_ int sd_pidfd_get_##name(int pidfd, char **ret) {       \
+                _cleanup_free_ char *str = NULL;                        \
+                pid_t pid;                                              \
+                int r;                                                  \
+                                                                        \
+                assert_return(pidfd >= 0, -EBADF);                      \
+                assert_return(ret, -EINVAL);                            \
+                                                                        \
+                r = pidfd_get_pid(pidfd, &pid);                         \
+                if (r < 0)                                              \
+                        return r;                                       \
+                                                                        \
+                r = sd_pid_get_##name(pid, &str);                       \
+                if (r < 0)                                              \
+                        return r;                                       \
+                                                                        \
+                r = pidfd_verify_pid(pidfd, pid);                       \
+                if (r < 0)                                              \
+                        return r;                                       \
+                                                                        \
+                *ret = TAKE_PTR(str);                                   \
+                return 0;                                               \
+        }
 
-        return 0;
-}
+DEFINE_PIDFD_GETTER(session);
+DEFINE_PIDFD_GETTER(unit);
+DEFINE_PIDFD_GETTER(user_unit);
+DEFINE_PIDFD_GETTER(machine_name);
+DEFINE_PIDFD_GETTER(slice);
+DEFINE_PIDFD_GETTER(user_slice);
+DEFINE_PIDFD_GETTER(cgroup);
```

Comment:
similar as above, i'd probably avoid the macros, and use generic functions taking function pointers.